vortex-array/src/variants.rs

```diff
@@ -7,126 +7,195 @@ use std::sync::Arc;
 
 use vortex_dtype::field::Field;
 use vortex_dtype::{DType, ExtDType, FieldNames, PType};
-use vortex_error::{vortex_panic, VortexExpect as _, VortexResult};
+use vortex_error::{vortex_panic, VortexError, VortexExpect as _, VortexResult};
 
-use crate::{ArrayData, ArrayTrait};
+use crate::encoding::Encoding;
+use crate::{ArrayDType, ArrayData, ArrayTrait};
 
-pub trait ArrayVariants {
-    fn as_null_array(&self) -> Option<&dyn NullArrayTrait> {
+/// An Array encoding must declare which DTypes it can be downcast into.
+pub trait VariantsVTable<Array> {
```

Comment:
nit - personally find real words for trait bounds very confusing

Reply:
Ha, this is the same comment @robert3005 made on the earlier VTable PRs (btw, all VTables use this convention).